### PR TITLE
Reduce interface terminal TPS impact + break some backends + new gui

### PR DESCRIPTION
**Pull request by firenoo**
~~I really wished the star import refactor was done after this, because now there are merge conflicts...~~ eh wasnt too bad, fixed

This is actually a subset of the features I want to implement.

Backend Changes
* Reduce the rate at which server-side changes are polled. Before this was done on every tick, but this has been changed to when changes were detected on either the ME network (i.e. a connection was added) or when the player performs an inventory action.
* Deprecate (it's only been like 5 versions lol) the old interface terminal handler. Despite approving the old one, I feel like it could have been done a bit better. This is a **breaking change** with GT5 and AE2FC.

GUI Changes
* Revisited the GUI. I felt like learning OpenGL, and decided to implement a true viewport widget for the interface terminal.
* Names of each interface title are pinned at the top.
* Shift + scroll = scroll from section to section
* Ctrl + scroll = go to top/bottom of the page.

Would like some brave testers to test this. Initial tests seemed fine, but there are probably a ton of corner cases that need to have bugs squashed.

### CODE REVIEW

File: src/main/java/appeng/client/render/TranslatedRenderItem.java
Comment (Laiff) [inline src/main/java/appeng/client/render/TranslatedRenderItem.java:26]: Can you give ability to configure position of the text or direct control on applied offsets?

It can be helpful to me to use in Level Maintainer terminal because I need to show at least two numbers for each stack.

File: src/main/java/appeng/core/features/registries/IfaceTermRegistry.java
Comment (Laiff) [inline src/main/java/appeng/core/features/registries/IfaceTermRegistry.java:12]: May be it would be better to not use shortcuts in name `InterfaceTerminalRegistry` it's easy to understand the code

File: src/main/java/appeng/client/render/TranslatedRenderItem.java
Comment (firenoo) [reply src/main/java/appeng/client/render/TranslatedRenderItem.java:26]: I don't think this is the right render item you want. This renders overlay at full scale. you probably want `AppEngRenderItem`

File: src/main/java/appeng/core/features/registries/IfaceTermRegistry.java
Comment (firenoo) [reply src/main/java/appeng/core/features/registries/IfaceTermRegistry.java:12]: I have it commented in the external interface. The whole point of the refactor was to not have long names and iface is clear imo

Comment (Laiff) [reply src/main/java/appeng/core/features/registries/IfaceTermRegistry.java:12]: dbtble point imo

File: src/main/java/appeng/client/render/TranslatedRenderItem.java
Comment (Laiff) [reply src/main/java/appeng/client/render/TranslatedRenderItem.java:26]: For better understanding, since I am not very strong in native GL graphics. How does it become clear that it is used for non-scaled text?

I did something similar for my terminal, but I haven’t yet found where exactly the text size is set.

Comment (firenoo) [reply src/main/java/appeng/client/render/TranslatedRenderItem.java:26]: Text size is set by glScale. Mc renders all text the same size otherwise

File: src/main/java/appeng/core/features/registries/IfaceTermRegistry.java
Comment (AnrDaemon) [reply src/main/java/appeng/core/features/registries/IfaceTermRegistry.java:12]: You are using IDE completion anyway.
Trying to "not use long names" often lead to comical situations.

File: src/main/java/appeng/core/sync/packets/PacketIfaceTermUpdate.java
Comment (Laiff) [inline src/main/java/appeng/core/sync/packets/PacketIfaceTermUpdate.java:63]: `AELog.error(...)` preferable to use here

File: src/main/java/appeng/core/sync/packets/PacketIfaceTermUpdate.java
Comment (Laiff) [inline src/main/java/appeng/core/sync/packets/PacketIfaceTermUpdate.java:56]: Flat magic numbers here can be replaced with enum match, from converted `int type` into PacketType

File: src/main/java/appeng/client/gui/implementations/GuiInterfaceTerminal.java
Comment (Laiff) [inline src/main/java/appeng/client/gui/implementations/GuiInterfaceTerminal.java:1]: Also this file is to big to show in diff and to big to understand code inside. 

May be it can be better if you split all inner classes into separate files inside new package `implementations.interfaceterminal` for example. That can help to improve readability and navigation thru the code

File: src/main/java/appeng/core/features/registries/IfaceTermRegistry.java
Comment (firenoo) [reply src/main/java/appeng/core/features/registries/IfaceTermRegistry.java:12]: What comical situation is happening here? It's not like a player will see this anyways. And even so, the opportunity to put "comical" names in otherwise boring code is cool. not like this a professional project

I am replacing `InterfaceTerminalSupportedClassProvider`. I've used the term `Iface` / `IfaceTerm` consistently in the new code with some documentation on how it's used, so still I don't see this as a strong reason to switch over

File: src/main/java/appeng/client/gui/implementations/GuiInterfaceTerminal.java
Comment (firenoo) [reply src/main/java/appeng/client/gui/implementations/GuiInterfaceTerminal.java:1]: When I wrote this, I considered performance first and foremost. Consider using github's commit-by-commit review feature as I've already put in the effort to make a clean commit history

These are non-static classes and I'd prefer to keep it that way over using references to the gui container instance + scope issues

File: src/main/java/appeng/core/features/registries/IfaceTermRegistry.java
Comment (Laiff) [reply src/main/java/appeng/core/features/registries/IfaceTermRegistry.java:15]: Here is points to not use shortened names
 - I know noting about code but name. An if I try to search something i'll write like `InterfaceTerminal` -> nothing found
 - Consistent typos because `Iface` word not present in any dictionary as a valid word
 - Knowledge about why this thing named with this strange composition of letters persisted only in one head

Especially `setDispRep` and `setSelfRep` is unreadable at all. I need to investigate whole data flow inside this variable to realize what they really means -> `setMachineIcon` and `setCraftableIcon`.

**Comment by Laiff** [comment]
What about realtime update in interface terminal from other interactions. Like other player put new pattern if interface or other path that valuable in my implementation for Level terminal state has been changed and we need to notify opened gui about this

File: src/main/java/appeng/core/sync/packets/PacketIfaceTermUpdate.java
Comment (Laiff) [inline src/main/java/appeng/core/sync/packets/PacketIfaceTermUpdate.java:252]: I hadn't where this buffer gracefully released. Could you check it for potential leakage

File: src/main/resources/assets/appliedenergistics2/textures/guis/newinterfaceterminal.png
Comment (Laiff) [inline src/main/resources/assets/appliedenergistics2/textures/guis/newinterfaceterminal.png:1]: Left side of viewport area is not properly aligned. Left border should be moved 3px to left

File: src/main/java/appeng/core/sync/packets/PacketIfaceTermUpdate.java
Comment (firenoo) [reply src/main/java/appeng/core/sync/packets/PacketIfaceTermUpdate.java:252]: I don't remember why this is direct in the first place, should be a heap buffer

Comment (Laiff) [reply src/main/java/appeng/core/sync/packets/PacketIfaceTermUpdate.java:252]: Ok better to wrap this in try block and gracefully release buffer in finally section

File: src/main/java/appeng/core/features/registries/IfaceTermRegistry.java
Comment (firenoo) [reply src/main/java/appeng/core/features/registries/IfaceTermRegistry.java:15]: Seems like I had comments on the client side but not the packet. that will be fixed, but currently it is unused by the interface  terminal.

You may feel free to call out missing documentation, but I will not be changing the name

File: src/main/java/appeng/core/sync/packets/PacketIfaceTermUpdate.java
Comment (firenoo) [reply src/main/java/appeng/core/sync/packets/PacketIfaceTermUpdate.java:252]: heap buffer should be collected by GC already

Comment (Laiff) [reply src/main/java/appeng/core/sync/packets/PacketIfaceTermUpdate.java]: Yes it can be collected by GC but here you put resource allocation on it and buffer should be released it means ref count should be 0 before it can be correctly pass GC phase

File: src/main/resources/assets/appliedenergistics2/textures/guis/newinterfaceterminal.png
Comment (firenoo) [reply src/main/resources/assets/appliedenergistics2/textures/guis/newinterfaceterminal.png:1]: That doesn't seem right, the left pixel is at 10px

File: src/main/java/appeng/core/sync/packets/PacketIfaceTermUpdate.java
Comment (firenoo) [reply src/main/java/appeng/core/sync/packets/PacketIfaceTermUpdate.java]: I don't think its needed but sure, I added a release
See https://github.com/netty/netty/issues/9991

**Comment by firenoo** [comment]
> What about realtime update in interface terminal from other interactions. Like other player put new pattern if interface or other path that valuable in my implementation for Level terminal state has been changed and we need to notify opened gui about this

I haven't tested this but one way you can do is inherit off the packet and put custom handling code to perform the proper GUI check

**Comment by Laiff** [comment]
An bug was detected with adding a duplicate template to the interface via the terminal.
After clicking on the same interface, which already has the same pattern, it disappears from your hands. But now 2 identical patterns are stored in one slot and you can take them one at a time

File: src/main/java/appeng/client/gui/implementations/GuiInterfaceTerminal.java
Comment (Laiff) [inline src/main/java/appeng/client/gui/implementations/GuiInterfaceTerminal.java:461]: Correct if I'm wrong but I think behaviour with green hilighter for founded patterns lost in new implementation

File: src/main/resources/assets/appliedenergistics2/textures/guis/newinterfaceterminal.png
Comment (Laiff) [reply src/main/resources/assets/appliedenergistics2/textures/guis/newinterfaceterminal.png:1]: But as I can see on other interfaces and here on other side padding is smaller than here on left side

Comment (firenoo) [reply src/main/resources/assets/appliedenergistics2/textures/guis/newinterfaceterminal.png:1]: I don't understand the problem, send me a screenshot of the misconfigured viewport

**Comment by Dream-Master** [comment]
@Laiff

**Comment by Laiff** [comment]
Yep lgtm, except this [one](https://github.com/GTNewHorizons/Applied-Energistics-2-Unofficial/pull/394#issuecomment-1751780443), already adapt some features from this PR into ae2fc. But my part isn't ready yet, but I can submit a draft into ae2fc

File: src/main/resources/assets/appliedenergistics2/textures/guis/newinterfaceterminal.png
Comment (Laiff) [reply src/main/resources/assets/appliedenergistics2/textures/guis/newinterfaceterminal.png:1]: pixil-frame-0 (1)
Aligned left side. 

pixil-frame-0
Problem highlithed

Comment (firenoo) [reply src/main/resources/assets/appliedenergistics2/textures/guis/newinterfaceterminal.png:1]: That was always like that... it was never symmetrical to begin with

**Review by Laiff — APPROVED** [review]

**Review by Laiff — APPROVED** [review]

File: src/main/java/appeng/container/implementations/ContainerInterfaceTerminal.java
Comment (firenoo) [inline src/main/java/appeng/container/implementations/ContainerInterfaceTerminal.java:64]: Where is this being used?

File: src/main/java/appeng/client/gui/IInterfaceTerminalPostUpdate.java
Comment (firenoo) [inline src/main/java/appeng/client/gui/IInterfaceTerminalPostUpdate.java:15]: Not so sure about creating a whole new interface for this. Does ae2fc not inherit off GuiInterfaceTerminal already to achieve the same effect

File: src/main/java/appeng/core/sync/packets/PacketIfaceTermUpdate.java
Comment (firenoo) [inline src/main/java/appeng/core/sync/packets/PacketIfaceTermUpdate.java:292]: Can this be a byte?

File: src/main/java/appeng/container/implementations/ContainerInterfaceTerminal.java
Comment (Laiff) [reply src/main/java/appeng/container/implementations/ContainerInterfaceTerminal.java:64]: InterfaceWireless container use it as delegate container

File: src/main/java/appeng/client/gui/IInterfaceTerminalPostUpdate.java
Comment (Laiff) [reply src/main/java/appeng/client/gui/IInterfaceTerminalPostUpdate.java:15]: Yes it implement same interface and packet can be delivered without any dirty hacks

File: src/main/java/appeng/core/sync/packets/PacketIfaceTermUpdate.java
Comment (Laiff) [reply src/main/java/appeng/core/sync/packets/PacketIfaceTermUpdate.java]: Yes it can. will fix it

**Comment by firenoo** [comment]
I will assign laiff so he will decide when to merge. I dont see too many open issues after the last round of fixes so I assume most of them are good. I would recommend another round of tests for each part of the terminal before merging.

File: src/main/java/appeng/client/gui/implementations/GuiInterfaceTerminal.java
Comment (miozune) [inline src/main/java/appeng/client/gui/implementations/GuiInterfaceTerminal.java:724]: nit: leftover

File: src/main/java/appeng/core/sync/packets/PacketCompressedNBT.java
Comment (miozune) [inline src/main/java/appeng/core/sync/packets/PacketCompressedNBT.java:91]: nit: leftover

**Review by miozune — COMMENTED** [review]
Textbox rendering looks broken in dev env

**Comment by Laiff** [comment]
>  Textbox rendering looks broken in dev env

What language should be here? I think utf flag somehow lost for this textboxes

**Comment by miozune** [comment]
I typed "yellow"

**Comment by Laiff** [comment]
> I typed "yellow"

ok i'll check it

**Comment by Laiff** [comment]
> I typed "yellow"
I can't reproduce this behaviour. I tried to enter this in plain english in japan and greek. There are no problem with rendering. Can you clarify what exactly you do to get this strange rendering?

**Review by miozune — APPROVED** [review]
The issue I raised is
- Cannot be reproduced on Laiff's env
- Cannot be reproduced on my modpack instance
- Cannot be figured out how to fix on my end

So I'd approve this PR. Great work!

**Comment by OneEyeMaker** [comment]
@Laiff Look carefully at Miozune's image. It's not the text rendering that's broken; it's the rendering order. Look at counts of items. These numbers are rendered *behind* items. Also word 'yellow' is correctly rendered  (if you check width of characters), but every character have some rectangular overlay.

**Comment by Laiff** [comment]
> Look at counts of items

Yes there are few problems fix zIndexes, mainly all of them related to DuraDisplay where Depth test completely disabled for overlays. I'll try to figure out solution but separately after discussing with caedis or someone else

**Pull request merged**
